// pages/editArticle/index.js
import React, { useEffect, useState, useContext } from 'react';

import ArticleForm from 'components/articleForm';
import useFetch from 'hooks/useFetch';
import { Redirect } from 'react-router';
import {CurrentUserContext} from 'contexts/currentUser';

const EditArticle = ({match}) => {
    const slug = match.params.slug;
    console.log('slug11', slug);
    const [currentUserState] = useContext(CurrentUserContext);
    const apiUrl = `/article/${slug}`;    // articles ?
    const [{response: fetchArticleResponse}, doFetchArticle] = useFetch(apiUrl);
    const [
        {response: updateArticleResponse, error: updateArticleError}, doUpdateArticle
    ] = useFetch(apiUrl);
    const [initialValues, setinitialValues] = useState(null);
    const [isSuccessfullSubmit, setIsSuccessfullSubmit] = useState(false);

    const handleSubmit = article => {
        console.log('handleSubmit', article)
        doUpdateArticle({
            metchod: 'put',
            data: {
                article
            }
        })
    }

    useEffect(() => {
        doFetchArticle()
    }, [doFetchArticle]);

    useEffect(() => {
        if(!fetchArticleResponse) {
            return
        }
        setinitialValues({
            title: fetchArticleResponse.article.title,
            description: fetchArticleResponse.article.description,
            body: fetchArticleResponse.article.body,
            tagList: fetchArticleResponse.article.tagList,
        })
    }, [fetchArticleResponse]);

    useEffect(() => {
        if (!updateArticleResponse) {
            return
        }
        setIsSuccessfullSubmit(true);
    }, [updateArticleResponse])

    if (currentUserState.isLoggerIn === false) {
        return <Redirect to={'/'} />
    }

    if (isSuccessfullSubmit) {
        return <Redirect to={`/articles/${slug}`} />
    }


    return (
        <ArticleForm 
            onSubmit={handleSubmit}
            errors={(updateArticleError && updateArticleError.errors) || {}}
            initialValues={initialValues}
        />
    )
};

export default EditArticle;
